Menus/PieMenuGUI.cpp

```diff
@@ -1,1416 +1,634 @@
-            // Left side, so align text right
-            else
-                pFont->DrawAligned(&allegroBitmap, drawPos.m_X + textPos.m_X, drawPos.m_Y + textPos.m_Y, m_pHoveredSlice->m_Description.c_str(), GUIFont::Right);
-        }
-    }
-/*
-    Vector mouseDir(m_InnerRadius * m_pController->GetAnalogCursor().GetMagnitude(), 0);
-    mouseDir.RadRotate(m_CursorAng);
-    line(pTargetBitmap, drawPos.GetFloorIntX(), drawPos.GetFloorIntY(), drawPos.GetFloorIntX() + mouseDir.m_X, drawPos.GetFloorIntY() + mouseDir.m_Y, 122);
-*/
-}
 
+		m_HoveredSlice = pieSliceToSelect;
+
+		m_BGBitmapNeedsRedrawing = true;
+		if (moveCursorToSlice) { m_CursorAngle = m_HoveredSlice->GetMidAngle(); }
+
+		SoundContainer *soundToPlay = pieSliceToSelect->IsEnabled() ? g_GUISound.HoverChangeSound() : g_GUISound.HoverDisabledSound();
+		soundToPlay->Play();
+		return true;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	//TODO Need to investigate this stuff, I don't fully understand what it does and maybe it's mostly undesirable since we'll want to be able to draw pie menus partly off the screen
+	void PieMenuGUI::CalculateDrawPositionAccountingForSeamsAndFont(const BITMAP *targetBitmap, const Vector &targetPos, Vector &drawPos) const {
+		drawPos = m_CenterPos - targetPos;
+		if (!targetPos.IsZero()) {
+			const Box *nearestBox = nullptr;
+
+			Box screenBox(targetPos, static_cast<float>(targetBitmap->w), static_cast<float>(targetBitmap->h));
+			list<Box> wrappedBoxes;
+			bool withinAnyBox = false;
+			float distance = std::numeric_limits<float>::max();
+			float shortestDist = std::numeric_limits<float>::max();
+			//TODO under what conditions would the pie menu not be on the screen and, if that's the case, would we still want to draw it? Investigate this!
+			g_SceneMan.WrapBox(screenBox, wrappedBoxes);
+			for (const Box &wrappedBox : wrappedBoxes) {
+				if (wrappedBox.IsWithinBox(m_CenterPos)) {
+					nearestBox = &wrappedBox;
+					withinAnyBox = true;
+					break;
+				}
+
+				distance = g_SceneMan.ShortestDistance(wrappedBox.GetCenter(), m_CenterPos).GetLargest();
+				if (distance < shortestDist) {
+					shortestDist = distance;
+					nearestBox = &wrappedBox;
+				}
+			}
+			drawPos += screenBox.GetCorner() - nearestBox->GetCorner();
+
+			if (!withinAnyBox) {
+				drawPos = nearestBox->GetCenter() + g_SceneMan.ShortestDistance(nearestBox->GetCenter(), m_CenterPos) - targetPos;
+			}
+		}
 
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          SelectSlice
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Sets a slice to be selected.
+		//TODO judging by the comment this will force it onto the screen if it's not on it? Better behaviour would be only draw if it's on the screen. See above TODO as well.
```

Comment:
My line of questioning how to handle pie menus that are off of the screen has to do with future strategic mode stuff.
Strategic mode will let you open pie menus at your mouse position, and there'll be cases where it'll be helpful to snap to targets. E.g. if you open a pie menu that has to do with an enemy door, you'll want to snap the pie menu to the middle of the door, even if the mouse cursor is a few pixels away.
Given this, especially with bigger actors and stuff, there could be cases where your pie menu is partly offscreen and probably shouldn't be forced onto screen. Though now I think about it, maybe the best solution to this would be to scroll the screen as needed to get the pie menu fully on it. That'll probably look/feel nice and will sidestep the issue.

Reply:
> Given this, especially with bigger actors and stuff, there could be cases where your pie menu is partly offscreen and probably shouldn't be forced onto screen.

What good does it do if you can't see the slices that are off screen and have to guess them?

>Though now I think about it, maybe the best solution to this would be to scroll the screen as needed to get the pie menu fully on it. That'll probably look/feel nice and will sidestep the issue.

This is a good enough solution.

Reply:
Menus being partly off screen would mean you need to scroll your view a bit to use em. But yeah, autoscrolling the screen would be better for sure. AndOff screen != off map. That's a good point though, even if pie menus can go off screen they mustn't be allowed to go off map on non-wrapping maps.
Anyway, this is all future shit, I'll leave this all as-is for now.